cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -0,0 +1,298 @@
+import com.google.common.util.concurrent.AbstractExecutionThreadService;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.common.util.concurrent.AbstractService;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.google.inject.Inject;
+import org.apache.tephra.RetryStrategies;
+import org.apache.tephra.TransactionFailureException;
+import org.apache.tephra.TransactionSystemClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Polls the JobQueue, checks the jobs for constraint satisfaction, and launches them.
+ */
+class ConstraintCheckerService extends AbstractIdleService {
+  private static final Logger LOG = LoggerFactory.getLogger(ConstraintCheckerService.class);
+
+  private final Transactional transactional;
+  private final DatasetFramework datasetFramework;
+  private final MultiThreadDatasetCache multiThreadDatasetCache;
+  private final Store store;
+  private final ProgramLifecycleService lifecycleService;
+  private final PropertiesResolver propertiesResolver;
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+  private final CConfiguration cConf;
+  private ScheduleTaskRunner taskRunner;
+  private ListeningExecutorService taskExecutorService;
+  private volatile boolean stopping = false;
+
+  @Inject
+  ConstraintCheckerService(Store store,
+                           ProgramLifecycleService lifecycleService, PropertiesResolver propertiesResolver,
+                           NamespaceQueryAdmin namespaceQueryAdmin,
+                           CConfiguration cConf,
+                           DatasetFramework datasetFramework,
+                           TransactionSystemClient txClient) {
+    this.store = store;
+    this.lifecycleService = lifecycleService;
+    this.propertiesResolver = propertiesResolver;
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+    this.cConf = cConf;
+    this.multiThreadDatasetCache = new MultiThreadDatasetCache(
+      new SystemDatasetInstantiator(datasetFramework), txClient,
+      NamespaceId.SYSTEM, ImmutableMap.<String, String>of(), null, null);
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(multiThreadDatasetCache),
+      RetryStrategies.retryOnConflict(20, 100)
+    );
+    this.datasetFramework = datasetFramework;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting ConstraintCheckerService.");
+    taskExecutorService = MoreExecutors.listeningDecorator(
+      Executors.newCachedThreadPool(new ThreadFactoryBuilder().setNameFormat("constraint-checker-task").build()));
```

Comment:
Thread names don't have to be unique. There can be multiple threads created in the thread pool, each with the same name.